### ISSUE TITLE
Mongo Sub entity with embedded documents

### ISSUE THREAD

**Issue by dudilo** [allcount/allcountjs#21]
Embedded doc should be added so a document could have many embedded documents. An example would be a project management application where a project is a doument and the assigned tasks are sub entities or as said embedded documents.

**Comment by AlexYegupov** [comment]
Embedded documents are the core of mongo. The implementation of this issue would be the very important step forward =)

**Comment by paveltiunov** [comment]
Yep. Agree. But we're building AllcountJS to be db agnostic and it's important to decide how to support this type of relation for SQL databases or just end up only with MongoDB support. Any thoughts on it?

**Comment by AlexYegupov** [comment]
Imagine some e-shop purchase order.

Typical USEFUL (in case of purchase order I strongly beleive user should be allowed to see order and detail grid at one page) user interface would be smth like that:


Order edit form
==============================================
<customer information controls>
...

<delivery information controls>
...


Order details grid
---------------------
 [product1]       [price]              [amount]    
 [product2]       [price]              [amount]    
 ...
=============================================


In case of mongo most probably right solution would be to impolement that as single document of "Orders" collection with several nested documents containing ordering products details, customer data, delivery information etc.
Hense admin configuration should be configured something like "use Order collection's fields ... to be modified on that form" and "use those Order's embedded document's (array) to be reflected at Order details grid control".


In case of relational db that purchase order most probably would consist of several db records with FK relations (Order, OrderItem, Customer, etc).
So that admin configuration in that case should has settings like "display fields of Customer table where id = Order.customer_id", "display fields of DeliveryInfo table where id = Order.delivery_id", "display all data of OrderDetails in order details grid". 

Aren't last two paragraphes semms too complicated to implement? And there were wrote nothing about UPDATING this form data. That operation would be very different for single document of mongo and several connected tables of relational db.
It's very hard for me to beleive that all above described shit could be implemented really elegant way for different db engines without either serious db-limitations like "not use embedded documents for mongo" (that's ideologically wrong in mongo) or terrible-simplifying user interface (like "display controls XOR grid but not together").

The ways that seems to me as possible to be implemented are:
1) single-db interface.
2) use user-specified REST-api server for db-read/update operations. The ng-admin engine goes that way. The cons of that solution seems like:
 - harder configuration 
 - slower speed
 - very limited interface form control options.

I didn't found good admin solutions for mongo (probably you too because you've implemented allcountjs) so that niche seems still vacant. It seems to me that relationships of user interface and db-engine on practice are really higher than we can imagine and the right way to make non-overcomplicated underhood is to use for mostry auto-generated admin interface single db engine.

**Comment by paveltiunov** [comment]
@AlexYegupov wow! Really great write up! 
You're totally right it's not simple to implement this sort of stuff when you need to support both MongoDB and SQL. But AllcountJS philosophy is all about that. We're building a bridge between Domain Model (Entities, Fields, ACL) and Technologies (UI, DB, etc.). We recently discussed with our team the way we should approach such things like this issue. You could take a look at this commit for example: https://github.com/allcount/allcountjs/commit/438a52e6f3a0b2b266b540a6ea4e9cb44d082026
From configuration standpoint I think it should look like:
```
Fields.relation(name, relationEntityTypeId, backReferenceField).embedded(true)
```
To implement different behavior we can use different implementations for mongooseModelCompileService, sqlModelCompileService and in storageDriver itself of course as in commit I mentioned. Our goal is to make AllcountJS core lean and make all other stuff as modules.

If you're concerned about request round-trip count then I think we should have an opportunity to load all relations in one round-trip. It doesn't matter is it embedded or not. Then it's probably another one issue.

What do you think?
